packages/enclave-contracts/contracts/token/EnclaveToken.sol

```diff
@@ -0,0 +1,233 @@
+    mapping(address account => bool allowed) public transferWhitelisted;
+
+    /// @dev Whether transfers are currently restricted.
+    bool public transfersRestricted;
+
+    /// @dev Emitted when tokens are minted as part of an allocation.
+    event AllocationMinted(
+        address indexed recipient,
+        uint256 amount,
+        string allocation
+    );
+
+    /// @dev Emitted whenever the transfer restriction flag is updated.
+    event TransferRestrictionUpdated(bool restricted);
+
+    /// @dev Emitted when an address is added to or removed from the whitelist.
+    event TransferWhitelistUpdated(address indexed account, bool whitelisted);
+
+    /**
+     * @notice Deploy the Enclave token.
+     * @param _owner Address that will initially own the contract and have admin rights.
+     */
+    constructor(
+        address _owner
+    ) ERC20("Enclave", "ENCL") ERC20Permit("Enclave") Ownable(_owner) {
+        // Grant the deployer all admin roles.
+        _grantRole(DEFAULT_ADMIN_ROLE, _owner);
+        _grantRole(MINTER_ROLE, _owner);
+
+        // Initialise state variables.
+        totalMinted = 0;
+        transfersRestricted = true;
+        transferWhitelisted[_owner] = true;
+
+        emit TransferRestrictionUpdated(true);
+        emit TransferWhitelistUpdated(_owner, true);
+    }
+
+    /**
+     * @notice Mint an allocation of tokens to a recipient.
+     * @dev Only accounts with the MINTER_ROLE may call this function.
+     * @param recipient Address to receive the minted tokens.
+     * @param amount Amount of tokens to mint (18 decimals).
+     * @param allocation Description of the allocation for off-chain bookkeeping.
+     */
+    function mintAllocation(
+        address recipient,
+        uint256 amount,
+        string memory allocation
+    ) external onlyRole(MINTER_ROLE) {
+        if (recipient == address(0)) revert ZeroAddress();
+        if (amount == 0) revert ZeroAmount();
+        // Ensure we do not exceed the total supply.
+        if (totalMinted + amount > TOTAL_SUPPLY) revert ExceedsTotalSupply();
+
+        _mint(recipient, amount);
+        totalMinted += amount;
+        emit AllocationMinted(recipient, amount, allocation);
+    }
+
+    /**
+     * @notice Mint multiple allocations in a batch.
+     * @dev Only accounts with the MINTER_ROLE may call this function.
+     * @param recipients Array of addresses to receive tokens.
+     * @param amounts Corresponding array of amounts to mint.
+     * @param allocations Array of allocation descriptions.
+     */
+    function batchMintAllocations(
+        address[] memory recipients,
+        uint256[] memory amounts,
+        string[] memory allocations
+    ) external onlyRole(MINTER_ROLE) {
+        if (
+            recipients.length != amounts.length ||
+            amounts.length != allocations.length
+        ) revert ArrayLengthMismatch();
+
+        uint256 totalAmount = 0;
+        for (uint256 i = 0; i < amounts.length; i++) {
```

Comment:
IIRC it's cheaper to cache the length to a variable first and pass it to the loop description rather than passing x.length directly